my/core/query.py

```diff
@@ -0,0 +1,717 @@
+
+    # user must provide either a key or a where predicate
+    if where_function is not None:
+        if isinstance(obj, dict):
+            for k, v in obj.items():
+                if where_function(v):
+                    return lambda o: o.get(k, default)  # type: ignore[union-attr]
+        elif dataclasses.is_dataclass(obj):
+            for (field_name, _annotation) in obj.__annotations__.items():
+                if where_function(getattr(obj, field_name)):
+                    return lambda o: getattr(o, field_name, default)
+        elif is_namedtuple(obj):
+            assert hasattr(obj, '_fields'), "Could not find '_fields' on attribute which is assumed to be a NamedTuple"
+            for field_name in getattr(obj, '_fields'):
+                if where_function(getattr(obj, field_name)):
+                    return lambda o: getattr(o, field_name, default)
+        # try using inspect.getmembers (like 'dir()') even if the dataclass/NT checks failed,
+        # since the attribute one is searching for might be a @property
+        for k, v in inspect.getmembers(obj):
+            if where_function(v):
+                return lambda o: getattr(o, k, default)
+
+    if default is not None:
+        # warn here? it seems like you typically wouldn't want to just set the order by to
+        # the same value everywhere, but maybe you did this on purpose?
+        return lambda _o: default
+    elif force_unsortable:
+        # generate a dummy function which returns None
+        # this causes this type of object to be classified as an unsortable item
+        return lambda _o: None
+    else:
+        return None  # couldn't compute a OrderFunc for this class/instance
+
+
+def _drop_exceptions(itr: Iterator[ET]) -> Iterator[T]:
+    """Return non-errors from the iterable"""
+    for o in itr:
+        if isinstance(o, Exception):
+            continue
+        yield o
+
+
+def _raise_exceptions(itr: Iterable[ET]) -> Iterator[T]:
+    """Raise errors from the iterable, stops the select function"""
+    for o in itr:
+        if isinstance(o, Exception):
+            raise o
+        yield o
+
+
+# currently using the 'key set' as a proxy for 'this is the same type of thing'
+def _determine_order_by_value_key(obj_res: ET) -> Any:
+    """
+    Returns either the class, or a tuple of the dictionary keys
+    """
+    key = obj_res.__class__
+    if key == dict:
+        # assuming same keys signify same way to determine ordering
+        return tuple(obj_res.keys())  # type: ignore[union-attr]
+    return key
+
+
+def _drop_unsorted(itr: Iterator[ET], orderfunc: OrderFunc) -> Iterator[ET]:
+    for o in itr:
+        if isinstance(o, Unsortable):
+            continue
+        ordval = orderfunc(o)
+        if ordval is None:
+            continue
+        yield o
+
+
+# hmm... is there really not a stdlib/more_itertools function to split an iterable by a predicate?
+# similar to 'my.core.error.sort_res_by'?
```

Comment:
yep! `https://more-itertools.readthedocs.io/en/stable/api.html#more_itertools.partition`
Although I often do something like
```
it1, it2 = tee(it)
res1 = (i for i in it1 if     isinstance(i, Unsortable))
res2 = (i for i in it2 if not isinstance(i, Unsortable))
return (res1, res2)
```
this makes it mypy-friendly, and slightly more iterative. Although in theory this still can end up taking linear memory depending on the downstream code, but so would storing them in a list, or partition -- don't think there is any way around it unless you're assuming you can iterate input iterable multiple times.

Reply:
Ah, was searching for `split` on the more_itertools docs... thanks.

Its become a bit more complicated since, as it has to check for recursive `Unsortable` instances, so it wouldnt be as simple as the above anymore.

Also seeing as whenever this is called it also has to call `sorted` (which will definitely consume the iterator), there isn't much benefit to iterators

think I will just leave as it is